### PR TITLE
Abstraction over Workflow engines and Containerization platforms

### PR DESCRIPTION
**Pull request by marcel-vesely-kinit**
We created interfaces/abstractions over `ReanaService` (`WorkflowBaseEngine`), as well as over `DockerService` (`ContainerBasePlatform`) in order to easily swap out these services for other technologies if necessary. Since there's a high chance of having seperate workflow engines for individual computing backends (kubernetes VS slurm), we wanted to make the integration of other such engines as smooth as possible.

In this branch, we also decoupled the all image building and experiment running logic into three seperate classes:
- `ExperimentScheduling` -> represents the scheduler, the planner that chooses which experiments/images to execute, etc... This class is the only one out of these three that have an access to database and can make changes to it
- `ContainerBasePlatform` -> represents containerization platforms (e.g., Docker, Singularity,...), that is in charge of building and pushing an image onto a remote registry
- `WorkflowBaseEngine` -> represents workflow engines (e.g., REANA, Argo workflows, Cycl,...), that focuses on running and further managing of workflows.

For now, this implementation is based on the idea that only one type of containerization platform or workflow engine is used, but later on, we may need to extend it to support multiple instances of platforms or engines working in parallel that would be utilized based on the user preference of where to execute his experiment.

### CODE REVIEW

File: backend/app/services/container_platforms/base_platform.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod, abstractstaticmethod
+
+from app.models.experiment_template import ExperimentTemplate
+
+
+class ContainerBasePlatform(ABC):
```
Comment (andrejridzik) [inline backend/app/services/container_platforms/base_platform.py:8]: I would rename the class to `ContainerPlatformBase` and the whole module simply to `base`

File: backend/app/services/scheduling.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+from app.schemas.experiment_template import ExperimentTemplateId
+from app.schemas.states import RunState, TemplateState
+from app.services.container_platforms.base_platform import ContainerBasePlatform
+from app.services.workflow_engines.base_engine import (
+    WorkflowBaseEngine,
+    WorkflowConnectionExcpetion,
+)
+
+
+class ExperimentScheduling:
```
Comment (andrejridzik) [inline backend/app/services/scheduling.py:25]: I recommend rename to `ExperimentScheduler` and module to `experiment_scheduler.py`

File: frontend/src/environments/environment.development.ts [hunk 3]
```diff
@@ -1,8 +1,7 @@
 export const environment = {
     BACKEND_API_URL: 'http://localhost:8000/v1',
     DEFAULT_PAGE_SIZE: 10,
-    AIOD_KEYCLOAK_URL: 'http://ec2-3-79-133-150.eu-central-1.compute.amazonaws.com/aiod-auth',
-    // AIOD_KEYCLOAK_URL: 'https://aiod-dev.i3a.es/aiod-aunpmth',
+    AIOD_KEYCLOAK_URL: 'https://aiod-dev.i3a.es/aiod-auth',
```
Comment (andrejridzik) [inline frontend/src/environments/environment.development.ts:4]: Please revert these changes for now. A separate PR is on its way for the overall FE deployment

File: backend/app/services/workflow_engines/base_engine.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+from app.helpers import WorkflowState
+from app.models.experiment import Experiment
+from app.models.experiment_run import ExperimentRun
+
+
+class WorkflowConnectionExcpetion(Exception):
+    pass
+
+
+class WorkflowBaseEngine(ABC):
```
Comment (andrejridzik) [inline backend/app/services/workflow_engines/base_engine.py:14]: I would rename the class to `WorkflowEngineBase and the whole module simply to `base`

**Review by andrejridzik — CHANGES_REQUESTED** [review]
I think you also forgot to commit the `__init__.py` files for the newly created sub-packages.

**Review by andrejridzik — APPROVED** [review]

File: backend/app/helpers.py [hunk 5]
```diff
@@ -35,3 +42,20 @@ def __call__(self):
...
 
 aiod_client_wrapper = AIoDClientWrapper()
+
+
+def create_env_file(env_vars: dict[str, str], path: Path) -> None:
+    lines = [f"{k}={v}" for k, v in env_vars.items()]
+    path.write_text("\n".join(lines))
+
+
+def setup_logger(logger_name: str) -> logging.Logger:
```
Comment (andrejridzik) [inline backend/app/helpers.py:52]: I just noticed that this is actually never used.. 🤔

Comment (marcel-vesely-kinit) [reply backend/app/helpers.py:52]: Yep, I had some problems with properly creating multiple new loggers that would maintain the original uvicorn formatting, so I gave up on it for now... I guess I should at least put there TODO so that we won't forget about it
